### ISSUE TITLE
CSR parsing ends in 'PHP Fatal error' => ParserException

### ISSUE THREAD

**Issue by dol** [fgrosse/PHPASN1#57]
I try to parse a CSR from binary. This following example ends with the following fatal error:

```
PHP Fatal error:  Uncaught exception 'FG\ASN1\Exception\ParserException' with message 'ASN.1 Parser Exception at offset 544: Context-Specific explicitly tagged object [0] starting at offset 544 is longer than allowed in the outer tag' in /tmp/foo/vendor/fgrosse/phpasn1/lib/ASN1/ExplicitlyTaggedObject.php:97
Stack trace:
#0 /tmp/foo/vendor/fgrosse/phpasn1/lib/ASN1/Object.php(201): FG\ASN1\ExplicitlyTaggedObject::fromBinary('0??X0??????0W1?...', 555)
#1 /tmp/foo/vendor/fgrosse/phpasn1/lib/ASN1/Construct.php(149): FG\ASN1\Object::fromBinary('0??X0??????0W1?...', 555)
#2 /tmp/foo/vendor/fgrosse/phpasn1/lib/ASN1/Object.php(222): FG\ASN1\Construct::fromBinary('0??X0??????0W1?...', 555)
#3 /tmp/foo/vendor/fgrosse/phpasn1/lib/ASN1/Construct.php(149): FG\ASN1\Object::fromBinary('0??X0??????0W1?...', 555)
#4 /tmp/foo/vendor/fgrosse/phpasn1/lib/ASN1/Object.php(222): FG\ASN1\Construct::fromBinary('0??X0??????0W1?...', 555)
#5 /tmp/foo/test.php(50): FG\ASN1\Object::fromBinary('0??X0??????0W1?...')
#6 {main}
  thrown in /tmp/foo/vendor/fgrosse/phpasn1/lib/ASN1/ExplicitlyTaggedObject.php on line 97
```

```php
<?php

require_once __DIR__ . '/vendor/autoload.php';

use FG\ASN1\Object;

// Test vector CSR from https://github.com/pyca/cryptography/blob/b5bb49d115e9f325d90bb96aebc15e6a1ef30f0d/vectors/cryptography_vectors/x509/requests/dsa_sha1.pem
$origCsr = <<<'EOF'
# Use this command to check the if the ASN1 parsing of openssl
(cat <<CATEOF | openssl asn1parse -inform PEM
-----BEGIN CERTIFICATE REQUEST-----
MIICWDCCAhgCAQAwVzEYMBYGA1UEAwwPY3J5cHRvZ3JhcGh5LmlvMQ0wCwYDVQQK
DARQeUNBMQswCQYDVQQGEwJVUzEOMAwGA1UECAwFVGV4YXMxDzANBgNVBAcMBkF1
c3RpbjCCAbYwggErBgcqhkjOOAQBMIIBHgKBgQCNf628CeKEqvppFUzqJBdwBJCe
UZ+LNdaFzeW07NyVg+dNNwoPiK2pjwJvJ3Yvs9XaeDb5ht/Ns1ieW5Jb6hFN78A+
+B2uMMJLvG3z1YjpNCe7pkID1KWxaHsrXjtkPUxhSXb4n5WjjT5MiQZfupdRTCLF
Ctu/KJFjp0tUhZs1twIVAINd5WvQfPf4LiAy/niUmu0ReqLvAoGAH3F7Wgd4L8Lk
5o4xH+qRpU7dNrhqxjTRTwWmipfq6dLvMfse895Cw9EA35ymT1vcKux7/ftHTPgx
/qBYU7XgWfLSSYCgrEY/HoGK81I+PLeaOdRfqScxiXdShCRpz4VAsBSRAk6q+85g
GOih9GWMND9Lp8CyHlN2oh9L64SRlh4DgYQAAoGABxPwdkH2Npu1qVRSdKLUwBmY
Nn+zcbueE0NjY2cu1o+CF0wt4FyOg5vG3laN1QuijY2dhxlCOq7FVX3xDXc6si1t
Zcu4eASml7yP2WW5Uvn36FDt8TyKzbXXU7bRDlngtXMuPIK6+hQDQrxKO7oWvQaB
yKai27t+/mziuEY7FwugADAJBgcqhkjOOAQDAy8AMCwCFGHVjcAo0BEIGKfYF9dC
NXJ8Ss/fAhQJe1LhmOzpXeFyc/CpJN8jzp2BiA==
-----END CERTIFICATE REQUEST-----
CATEOF
)

EOF;

echo $origCsr;

$csr = <<<'EOF'
MIICWDCCAhgCAQAwVzEYMBYGA1UEAwwPY3J5cHRvZ3JhcGh5LmlvMQ0wCwYDVQQK
DARQeUNBMQswCQYDVQQGEwJVUzEOMAwGA1UECAwFVGV4YXMxDzANBgNVBAcMBkF1
c3RpbjCCAbYwggErBgcqhkjOOAQBMIIBHgKBgQCNf628CeKEqvppFUzqJBdwBJCe
UZ+LNdaFzeW07NyVg+dNNwoPiK2pjwJvJ3Yvs9XaeDb5ht/Ns1ieW5Jb6hFN78A+
+B2uMMJLvG3z1YjpNCe7pkID1KWxaHsrXjtkPUxhSXb4n5WjjT5MiQZfupdRTCLF
Ctu/KJFjp0tUhZs1twIVAINd5WvQfPf4LiAy/niUmu0ReqLvAoGAH3F7Wgd4L8Lk
5o4xH+qRpU7dNrhqxjTRTwWmipfq6dLvMfse895Cw9EA35ymT1vcKux7/ftHTPgx
/qBYU7XgWfLSSYCgrEY/HoGK81I+PLeaOdRfqScxiXdShCRpz4VAsBSRAk6q+85g
GOih9GWMND9Lp8CyHlN2oh9L64SRlh4DgYQAAoGABxPwdkH2Npu1qVRSdKLUwBmY
Nn+zcbueE0NjY2cu1o+CF0wt4FyOg5vG3laN1QuijY2dhxlCOq7FVX3xDXc6si1t
Zcu4eASml7yP2WW5Uvn36FDt8TyKzbXXU7bRDlngtXMuPIK6+hQDQrxKO7oWvQaB
yKai27t+/mziuEY7FwugADAJBgcqhkjOOAQDAy8AMCwCFGHVjcAo0BEIGKfYF9dC
NXJ8Ss/fAhQJe1LhmOzpXeFyc/CpJN8jzp2BiA==
EOF;

// $csr = str_replace("\n", '', $csr);
$asnObject = Object::fromBinary(base64_decode($csr));
```

Can you figure out what I'm doing wrong or do you spot the problem. Currently used version is 1.3.2 due to a local limitation of using PHP 5.6 requirement.

**Comment by fgrosse** [comment]
I just had a look at your issue and found a bug that prevents any tagged objects to be empty.
This will be fixed in 1.4.2 and for you on 1.3.3

**Comment by fgrosse** [comment]
If you reallly need to stick to an older PHP version you can use the [dev-compat/php5-5 branch alias][1] in your composer.json. For everybody else, this fix is avilable in [PHPASN version >= 1.4.2][2].

Thanks for reporting the issue Dominic :)

[1]: https://packagist.org/packages/fgrosse/phpasn1#dev-compat/php5-5
[2]: https://packagist.org/packages/fgrosse/phpasn1#1.4.2

**Comment by dol** [comment]
Thank you for the fix.

**Comment by dol** [comment]
I'd like to reopen this issue because I found some multi attribute CSR's that have the same parsing issue.

E.g:
```php
<?php
// https://github.com/phpseclib/phpseclib/blob/master/tests/Unit/File/X509/CSRTest.php
$csrWithMultipleAttributes = <<<'EOF'
MIIFGDCCAwACAQAwOjEWMBQGCgmSJomT8ixkARkWBnNlY3VyZTEgMB4GA1UEAxMX
LlNlY3VyZSBFbnRlcnByaXNlIENBIDEwggIiMA0GCSqGSIb3DQEBAQUAA4ICDwAw
ggIKAoICAQCzgEpL+Za7a3y7YpURDrxlGIBlks25fD0tHaZIYkBTaXA5h+9MWoXn
FA7AlIUt8pbBvXdJbOCmGaeQmBfBH0Qy9vTbx/DR2IOwzqy2ZHuurI5bPL12ceE2
Mxa9xgY/i7U6MAUtoA3amEd7cKj2fz9EWZruRladOX0DXv9KexSan+45QjCWH+u2
Cxem2zH9ZDNPGBuAF9YsAvkdHdAoX8aSm05ZAjUiO2e/+L57whh7zZiDY3WIhin7
N/2JNTKVO6lx50S8a34XUKBt3SKgSR941hcLrBYUNftUYsTPo40bzKKcWqemiH+w
jQiDrln4V2b5EbVeoGWe4UDPXCVmC6UPklG7iYfF0eeK4ujV8uc9PtV2LvGLOFdm
AYE3+FAba5byQATw/DY8EJKQ7ptPigJhVe47NNeJlsKwk1haJ9k8ZazjS+vT45B5
pqe0yBFAEon8TFnOLnAOblmKO12i0zqMUNAAlmr1c8jNjLr+dhruS+QropZmzZ24
mAnFG+Y0qpfhMzAxTGQyVjyGwDfRK/ARmtrGpmROjj5+6VuMmZ6Ljf3xN09epmtH
gJe+lYNBlpfUYg16tm+OusnziYnXL6nIo2ChOY/7GNJJif9fjvvaPDCC98K64av5
5rpIx7N/XH4hwHeQQkEQangExE+8UMyBNFNmvPnIHVHUZdYo4SLsYwIDAQABoIGY
MBsGCisGAQQBgjcNAgMxDRYLNi4zLjk2MDAuMi4weQYJKoZIhvcNAQkOMWwwajAQ
BgkrBgEEAYI3FQEEAwIBADAdBgNVHQ4EFgQU5nEIMEUT5mMd1WepmviwgK7dIzww
GQYJKwYBBAGCNxQCBAweCgBTAHUAYgBDAEEwCwYDVR0PBAQDAgGGMA8GA1UdEwEB
/wQFMAMBAf8wDQYJKoZIhvcNAQELBQADggIBAKZl6bAeaID3b/ic4aztL8ZZI7vi
D3A9otUKx6v1Xe63zDPR+DiWSnxb9m+l8OPtnWkcLkzEIM/IMWorHKUAJ/J871D0
Qx+0/HbkcrjMtVu/dNrtb9Z9CXup66ZvxTPcpEziq0/n2yw8QdBaa+lli65Qcwcy
tzMQK6WQTRYfvVCIX9AKcPKxwx1DLH+7hL/bERB1lUDu59Jx6fQfqJrFVOY2N8c0
MGvurfoHGmEoyCMIyvmIMu4+/wSNEE/sSDp4lZ6zuF6rf1m0GiLdTX2XJE+gfvep
JTFmp4S3WFqkszKvaxBIT+jV0XKTNDwnO+dpExwU4jZUh18CdEFkIUuQb0gFF8B7
WJFVpNdsRqZRPBz83BW1Kjo0yAmaoTrGNmG0p6Qf3K2zbk1+Jik3VZq4rvKoTi20
6RvLA2//cMNfkYPsuqvoHGe2e0GOLtIB63wJzloWROpb72ohEHsvCKullIJVSuiS
9sfTBAenHCyndgAEd4T3npTUdaiNumVEm5ilZId7LAYekJhkgFu3vlcl8blBJKjE
skVTp7JpBmdXCL/G/6H2SFjca4JMOAy3DxwlGdgneIaXazHs5nBK/BgKPIyPzZ4w
secxBTTCNgI48YezK3GDkn65cmlnkt6F6Mf0MwoDaXTuB88Jycbwb5ihKnHEJIsO
draiRBZruwMPwPIP
EOF;

// https://github.com/gknocke/Crypt-PKCS10/blob/master/t/00_base.t
$csrPerl = <<<'EOF'
MIIC4zCCAcsCAQAwcjELMAkGA1UEBhMCQVUxEzARBgNVBAgMClNvbWUtU3RhdGUx
ITAfBgNVBAoMGEludGVybmV0IFdpZGdpdHMgUHR5IEx0ZDENMAsGA1UEAwwEdGVz
dDEcMBoGCSqGSIb3DQEJARYNdGVzdEB0ZXN0LmNvbTCCASIwDQYJKoZIhvcNAQEB
BQADggEPADCCAQoCggEBAL7t6CR90kOZ0m3qJx9koHBFdXnunEq9ELSUvsZ7oFiw
6zJsyi18S9znVRLEgsW4v2rTsLiJmKnMF2Kyd9AQ2X2zKUK49CTMqG1p/4lF0oLa
nQgNmjYC4VL61wTte6CUAJX7wZA8LL00QfCdgILKodP12E3c7Yp3yw5TAOBdA3Av
KnpsUiNLmQN3HSn9dwk9rCHulrP8LL97l//H2KSCvOvSfk1XeN7ADSyWMC6ZPBD/
7WVnQo7GgMAA0Qhz+nIIFA5AyuFEbRiff+L5EEl+Bg6R9tgx+QUt9JXxF260dkAb
iKBmPcJldl2/GEmz6RAXhMwJnN35tbdzw0LVLzsypr0CAwEAAaAsMBMGCSqGSIb3
DQEJAjEGDAR0ZXN0MBUGCSqGSIb3DQEJBzEIDAYxMjM0NTYwDQYJKoZIhvcNAQEF
BQADggEBACqZ5+3GP7QbTJAqnUloMRBJGYZl0/EyW6dAMoq7E+RiR2OlJGXgP+ga
DPv7Wa4jZiYmf2CZCWKP+WvTvcg4AQq5h7hhAN+u+s8kk0lIjYjnkKRAedGYKDUO
fPD9WowISru3RB1xyxlvgeZS6WoA6TD8REVa1UyFoNzUewvsrNVkKSHh1xk/+ePx
2Ovvrcg9pAWY4K8FvMRdFQKnEud9CAoMxqz3kszhxDW6rcr2mgFPSrKi5WNj+Scg
Tqod8xbB753JWjEbG6Hui9LIMihTX3ZJ0c2GB0buhEgz49y83X/byFHSGGSQpzxX
qXDFVov9UZ+sGy8CJ5ahII79yrfKpxY=
EOF;

```

Online ASN1 parser view:
[csrWithMultipleAttributes](https://lapo.it/asn1js/#3082051830820300020100303A31163014060A0992268993F22C64011916067365637572653120301E060355040313172E53656375726520456E7465727072697365204341203130820222300D06092A864886F70D01010105000382020F003082020A0282020100B3804A4BF996BB6B7CBB6295110EBC6518806592CDB97C3D2D1DA64862405369703987EF4C5A85E7140EC094852DF296C1BD77496CE0A619A7909817C11F4432F6F4DBC7F0D1D883B0CEACB6647BAEAC8E5B3CBD7671E1363316BDC6063F8BB53A30052DA00DDA98477B70A8F67F3F44599AEE46569D397D035EFF4A7B149A9FEE394230961FEBB60B17A6DB31FD64334F181B8017D62C02F91D1DD0285FC6929B4E590235223B67BFF8BE7BC2187BCD98836375888629FB37FD893532953BA971E744BC6B7E1750A06DDD22A0491F78D6170BAC161435FB5462C4CFA38D1BCCA29C5AA7A6887FB08D0883AE59F85766F911B55EA0659EE140CF5C25660BA50F9251BB8987C5D1E78AE2E8D5F2E73D3ED5762EF18B385766018137F8501B6B96F24004F0FC363C109290EE9B4F8A026155EE3B34D78996C2B093585A27D93C65ACE34BEBD3E39079A6A7B4C811401289FC4C59CE2E700E6E598A3B5DA2D33A8C50D000966AF573C8CD8CBAFE761AEE4BE42BA29666CD9DB89809C51BE634AA97E13330314C6432563C86C037D12BF0119ADAC6A6644E8E3E7EE95B8C999E8B8DFDF1374F5EA66B478097BE9583419697D4620D7AB66F8EBAC9F38989D72FA9C8A360A1398FFB18D24989FF5F8EFBDA3C3082F7C2BAE1ABF9E6BA48C7B37F5C7E21C077904241106A7804C44FBC50CC81345366BCF9C81D51D465D628E122EC630203010001A08198301B060A2B0601040182370D0203310D160B362E332E393630302E322E307906092A864886F70D01090E316C306A301006092B06010401823715010403020100301D0603551D0E04160414E67108304513E6631DD567A99AF8B080AEDD233C301906092B0601040182371402040C1E0A00530075006200430041300B0603551D0F040403020186300F0603551D130101FF040530030101FF300D06092A864886F70D01010B05000382020100A665E9B01E6880F76FF89CE1ACED2FC65923BBE20F703DA2D50AC7ABF55DEEB7CC33D1F838964A7C5BF66FA5F0E3ED9D691C2E4CC420CFC8316A2B1CA50027F27CEF50F4431FB4FC76E472B8CCB55BBF74DAED6FD67D097BA9EBA66FC533DCA44CE2AB4FE7DB2C3C41D05A6BE9658BAE50730732B733102BA5904D161FBD50885FD00A70F2B1C31D432C7FBB84BFDB1110759540EEE7D271E9F41FA89AC554E63637C734306BEEADFA071A6128C82308CAF98832EE3EFF048D104FEC483A78959EB3B85EAB7F59B41A22DD4D7D97244FA07EF7A9253166A784B7585AA4B332AF6B10484FE8D5D17293343C273BE769131C14E23654875F02744164214B906F480517C07B589155A4D76C46A6513C1CFCDC15B52A3A34C8099AA13AC63661B4A7A41FDCADB36E4D7E262937559AB8AEF2A84E2DB4E91BCB036FFF70C35F9183ECBAABE81C67B67B418E2ED201EB7C09CE5A1644EA5BEF6A21107B2F08ABA59482554AE892F6C7D30407A71C2CA77600047784F79E94D475A88DBA65449B98A564877B2C061E909864805BB7BE5725F1B94124A8C4B24553A7B26906675708BFC6FFA1F64858DC6B824C380CB70F1C2519D8277886976B31ECE6704AFC180A3C8C8FCD9E30B1E7310534C2360238F187B32B7183927EB972696792DE85E8C7F4330A036974EE07CF09C9C6F06F98A12A71C4248B0E76B6A244166BBB030FC0F20F)
[csrPerl](// https://lapo.it/asn1js/#308202E3308201CB0201003072310B30090603550406130241553113301106035504080C0A536F6D652D53746174653121301F060355040A0C18496E7465726E6574205769646769747320507479204C7464310D300B06035504030C0474657374311C301A06092A864886F70D010901160D7465737440746573742E636F6D30820122300D06092A864886F70D01010105000382010F003082010A0282010100BEEDE8247DD24399D26DEA271F64A070457579EE9C4ABD10B494BEC67BA058B0EB326CCA2D7C4BDCE75512C482C5B8BF6AD3B0B88998A9CC1762B277D010D97DB32942B8F424CCA86D69FF8945D282DA9D080D9A3602E152FAD704ED7BA0940095FBC1903C2CBD3441F09D8082CAA1D3F5D84DDCED8A77CB0E5300E05D03702F2A7A6C52234B9903771D29FD77093DAC21EE96B3FC2CBF7B97FFC7D8A482BCEBD27E4D5778DEC00D2C96302E993C10FFED6567428EC680C000D10873FA7208140E40CAE1446D189F7FE2F910497E060E91F6D831F9052DF495F1176EB476401B88A0663DC265765DBF1849B3E9101784CC099CDDF9B5B773C342D52F3B32A6BD0203010001A02C301306092A864886F70D01090231060C0474657374301506092A864886F70D01090731080C06313233343536300D06092A864886F70D010105050003820101002A99E7EDC63FB41B4C902A9D4968311049198665D3F1325BA740328ABB13E4624763A52465E03FE81A0CFBFB59AE236626267F609909628FF96BD3BDC838010AB987B86100DFAEFACF249349488D88E790A44079D19828350E7CF0FD5A8C084ABBB7441D71CB196F81E652E96A00E930FC44455AD54C85A0DCD47B0BECACD5642921E1D7193FF9E3F1D8EBEFADC83DA40598E0AF05BCC45D1502A712E77D080A0CC6ACF792CCE1C435BAADCAF69A014F4AB2A2E56363F927204EAA1DF316C1EF9DC95A311B1BA1EE8BD2C83228535F7649D1CD860746EE844833E3DCBCDD7FDBC851D2186490A73C57A970C5568BFD519FAC1B2F022796A1208EFDCAB7CAA716)

**Comment by fgrosse** [comment]
The issue with the two given inputs is that the tag seems to be applied to multiple objects instead of exactly one. I didnt know this was possible and have to meditate over the solution a bit :)

**Comment by fgrosse** [comment]
The second fix is available in [version 1.5.0][1]. I decided for the minor version increase since having multiple objects under one tag felt more like a feature to me. Again I ported the change to the [dev-compat/php5-5 branch alias][2] as well.

Additionally you should have a look at [examples/Issue57.php][3] which contains all three of your inputs.

Best regards
Friedrich

[1]: https://packagist.org/packages/fgrosse/phpasn1#1.5.0
[2]: https://packagist.org/packages/fgrosse/phpasn1#dev-compat/php5-5
[3]: https://github.com/fgrosse/PHPASN1/blob/master/examples/Issue57.php

**Comment by dol** [comment]
Thank you.
